HDDS-9563. Containerize website development server

devabhishekpal opened the pull request:
This PR adds the required changes for creation of a docker file, which anybody can run with docker-compose and check changes on the go.
No package installations required.

#### What will this PR will do?
- It will build an image with the node_modules already installed in it.
- `docker-compose up` will create a container with the local changes mounted on this container.
- Site will be available to view on `localhost:3000` (default port mapping in the container)

#### What this means for the end-user?
- While the site can be deployed without an image needing to be built, as discussed in the [JIRA](https://issues.apache.org/jira/browse/HDDS-9563), we can make the developer work much smoother with a containerized environment
- Any new contributor can just simply spin up a container from the image, and make changes in their local development directory
- These changes will be picked up by the container in the mounted volume and live changes will be visible on localhost:3000

errose28 commented on compose.yml:
```diff
@@ -0,0 +1,28 @@
+      - "3000:3000"
+    volumes:
+      - ".:/ozone-site"
+      - /ozone-site/node_modules/
```
Let's add a comment explaining that this prevents the host's node_modules directory from being mounted if it exists, so that the one in the image is used.

errose28 replied:
This does end up creating an empty node_modules directory in the host, but I think that is an ok trade off.

devabhishekpal replied:
Added the comment to clarify the usage of the line, thanks for the suggestion.
Now that I look at it, it would have caused confusion down the line to others.

errose28 commented on compose.yml:
```diff
@@ -0,0 +1,28 @@
+      - ".:/ozone-site"
+      - /ozone-site/node_modules/
+    working_dir: "/ozone-site"
+    command: sh -c "pnpm start --host '0.0.0.0'"
```
```suggestion
    command: "pnpm start --host '0.0.0.0'"
```

For some reason this takes ~10 seconds for docker compose to stop if `sh` is used, but it stops instantly if `pnpm` is invoked directly. Something with the sigint/sigterm handling I think.

devabhishekpal replied:
Thanks for suggesting the improvement.
Solved in the latest commit

devabhishekpal replied:
Thanks for suggesting the improvement.
Solved in the latest commit

errose28 commented on README.md:
```diff
@@ -54,6 +54,24 @@ Once installed, there are two ways to preview the website locally. Each option w
+
+
+
+
```
I refactored the README now that we have the new setup option. Check it out [here](https://github.com/errose28/ozone-site/commit/06a68fcf28e4344e61160a643757ae485f779c43#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5) and if you think it looks good we can add that to this PR.

devabhishekpal replied:
Thanks, adding that version to the PR

errose28 reviewed (commented):
Thanks for taking this up and fixing the node issues @devabhishekpal. This is very satisfying to use 😄 I tested locally after removing my existing node installs and docker images and everything worked smoothly. I just left some comments inline for improvements.

Also can we add a `.dockerignore` file with the following contents to leave some unnecessary things out of the volume mount?
```
node_modules
build
.docusaurus
.cache-loader
.git
.gitignore
```
 It doesn't affect functionality, but if looking at the contents from within the container I think this makes it clearer what is actually being used in the preview. Especially if the host already has a build of the website present.

I have a testing branch [here](https://github.com/errose28/ozone-site/tree/HDDS-9563-containerize-website-test) that has these changes and it seems to work well.

devabhishekpal commented:
Thank you @errose28 for the review and for the improved ReadMe file reference.
The latest [commit](https://github.com/apache/ozone-site/pull/50/commits/1e14bfc30dd8883ca3b765450e433576025f3c43) addressed the review comments.

errose28 commented on compose.yml:
```diff
@@ -0,0 +1,31 @@
+      - 3000:3000
+    volumes:
+      - ".:/ozone-site"
+      # The below option is used to prevent overwriting of the docker image
+      # node_modules by the mounted volume's node_modules
+      # This would create an empty node_modules in the working directory of the local machine if not already present.
```
```suggestion
      # The below option is used to prevent overwriting the node_modules directory in the docker image
      # with the node_modules present in the mounted volume.
      # This will create an empty node_modules directory in the working directory of the host if it is not already present.
```

devabhishekpal replied:
Addressed in the latest commit

errose28 reviewed (approved):
One minor nit in the comment but otherwise LGTM. Thanks for the contribution @devabhishekpal.
